File: Class/Practise 2.1/hello.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>
#include <print>
#include <ranges>
#include <sstream>

using namespace std;

const string FILENAME = "user_stats.txt";

struct UserStat {
    string name;
    int count;

    UserStat(const string& name = "", int count = 0) : name(name), count(count) {}

    friend ostream& operator<<(ostream& os, const UserStat& stat) {
        os << stat.name << " " << stat.count;
        return os;
    }

    friend istream& operator>>(istream& is, UserStat& stat) {
        is >> stat.name >> stat.count;
        return is;
    }
};

class UserStatsManager {
private:
    unordered_map<string, UserStat> userStats;

public:
    UserStatsManager() {
        loadUserStats();
    }

    void loadUserStats() {
        ifstream file(FILENAME);

        if (file.is_open()) {
            string line;
            while (getline(file, line)) {
                istringstream iss(line);
                UserStat stat;
                iss >> stat;
                userStats[stat.name] = stat;
            }
            file.close();
        }
    }

    void saveUserStats() const {
        ofstream file(FILENAME, ios::trunc);

        if (file.is_open()) {
            for (const auto& entry : userStats | views::values) {
                file << entry << endl;
            }
            file.close();
        }
    }

    void greetUser(const string& name) {
        if (name == "bread") {
            userStats.clear();
            println("All user history has been exterminated!");
            return;
        }

        if (userStats.find(name) == userStats.end()) {
            println("Welcome, {}!", name);
            userStats[name] = UserStat(name, 1);
        } else {
            userStats[name].count++;
            println("Hello again (x{}), {}!", userStats[name].count, name);
        }
    }

    void resetUserStats(const string& name) {
        userStats.erase(name);
        println("Statistics for {} have been reset.", name);
    }

    unordered_map<string, UserStat>& getUserStats() {
        return userStats;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2 || argc > 3) {
        println("Usage: {} <name> [delete]", argv[0]);
        return 1;
    }

    string name = argv[1];
    bool deleteStats = (argc == 3 && string(argv[2]) == "delete");

    UserStatsManager manager;

    if (deleteStats) {
        manager.resetUserStats(name);
    } else {
        manager.greetUser(name);
    }

    manager.saveUserStats();

    return 0;
}

File: Home/Home 2.1/hello.swift
import Foundation

struct UserStat: Codable {
    var name: String
    var count: Int
}

class UserStatsManager {
    private var userStats: [String: UserStat] = [:]
    private let filename = "user_stats.txt"
    
    init() {
        loadUserStats()
    }
    
    private func loadUserStats() {
        if let data = try? String(contentsOfFile: filename) {
            data.split(separator: "\n").forEach { line in
                let parts = line.split(separator: " ")
                if parts.count == 2, let count = Int(parts[1]) {
                    let stat = UserStat(name: String(parts[0]), count: count)
                    userStats[stat.name] = stat
                }
            }
        }
    }
    
    private func saveUserStats() {
        let data = userStats.values.map { "\($0.name) \($0.count)" }.joined(separator: "\n")
        do {
            try data.write(toFile: filename, atomically: true, encoding: .utf8)
        } catch {
            print("Failed to write to file \(filename): \(error)")
        }
    }
    
    func greetUser(_ name: String) {
        if name == "bread" {
            userStats.removeAll()
            print("All user history has been exterminated!")
            saveUserStats()
            return
        }
        
        if userStats[name] == nil {
            print("Welcome, \(name)!")
            userStats[name] = UserStat(name: name, count: 1)
        } else {
            userStats[name]!.count += 1
            print("Hello again (x\(userStats[name]!.count)), \(name)!")
        }
        saveUserStats()
    }
    
    func resetUserStats(_ name: String) {
        userStats.removeValue(forKey: name)
        print("Statistics for \(name) have been reset.")
        saveUserStats()
    }
}

let arguments = CommandLine.arguments

guard arguments.count >= 2 && arguments.count <= 3 else {
    print("Usage: \(arguments[0]) <name> [delete]")
    exit(1)
}

let name = arguments[1]
let deleteStats = arguments.count == 3 && arguments[2] == "delete"

let manager = UserStatsManager()

if deleteStats {
    manager.resetUserStats(name)
} else {
    manager.greetUser(name)
}
